File: apps/ios/App/DashboardPrefs.swift
import Foundation
import SwiftUI

/**
 Which tiles are on the dashboard, in what order, and how wide each one is.

 A faithful port of `apps/web/src/dashboard.ts`, down to the two storage keys
 (`dashboardTiles`, `dashboardTileSizes`) and the JSON shapes, so a saved
 dashboard means the same thing wherever it is read. Same pattern as
 `NavPrefs` — an `ObservableObject` singleton over `UserDefaults`.

 **This is per-device and does not sync**, and that is web's behaviour too:
 `dashboard.ts` writes to localStorage, not to a table. Putting the native copy
 in the database "because we can" would make two devices disagree about a
 preference the browser has always kept to itself.

 Web stores a **third** key, `dashboardTileSpans`, plus `TileSpan`,
 `setTileSpan`, `useTileSpans`, `H_ROWS` and `isTileEnabled`. Every one of them
 is exported and unimported — a first sizing attempt that the `{w,h}` system
 replaced and nobody deleted. None of it is ported.
 */
@MainActor
final class DashboardPrefs: ObservableObject {
    static let shared = DashboardPrefs()

    /// Enabled tiles, in the user's chosen order.
    @Published private(set) var ids: [TileId]

    /// Per-tile width overrides. A tile with no entry uses ``defaultWidth``.
    @Published private(set) var widths: [TileId: TileWidth]

    /// The width a tile starts at.
    ///
    /// Web's grid is four columns wide and its default size is `md` (two of
    /// them) — half the row. An iPhone renders one column whatever this says,
    /// so the value only becomes visible on an iPad, which is exactly where a
    /// wrong default would look worst.
    static let defaultWidth: TileWidth = .md

    private let defaults: UserDefaults

    private init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        ids = DashboardPrefs.readIds(from: defaults)
        widths = DashboardPrefs.readWidths(from: defaults)
    }

    func width(of id: TileId) -> TileWidth { widths[id] ?? DashboardPrefs.defaultWidth }

    /// Add or remove a tile. Adding appends, which keeps a custom order intact.
    func setEnabled(_ id: TileId, _ on: Bool) {
        if on {
            guard !ids.contains(id) else { return }
            write(ids + [id])
        } else {
            write(ids.filter { $0 != id })
        }
    }

    /// Persist an explicit order — drag, or the move-up/move-down buttons.
    func reorder(_ ordered: [TileId]) {
        var seen = Set<TileId>()
        write(ordered.filter { seen.insert($0).inserted })
    }

    /// Move one tile by one position. Returns false at either end.
    @discardableResult
    func move(_ id: TileId, by delta: Int) -> Bool {
        guard let from = ids.firstIndex(of: id) else { return false }
        let to = from + delta
        guard to >= 0, to < ids.count else { return false }
        var next = ids
        next.remove(at: from)
        next.insert(id, at: to)
        write(next)
        return true
    }

    func setWidth(_ id: TileId, _ width: TileWidth) {
        var next = widths
        next[id] = width
        // Web stores {w,h}. Only `w` is ever read back — `H_ROWS` exists but
        // nothing uses it, and row height is measured from content instead. The
        // `h` is written so a dashboard saved on a phone still parses in the
        // browser, rather than tripping web's `isDim(s.h)` guard and being
        // dropped whole.
        let payload = next.reduce(into: [String: [String: String]]()) { acc, pair in
            acc[pair.key.rawValue] = ["w": pair.value.rawValue, "h": "md"]
        }
        if let data = try? JSONSerialization.data(withJSONObject: payload),
           let json = String(data: data, encoding: .utf8) {
            defaults.set(json, forKey: tileSizeKey)
        }
        widths = next
    }

    private func write(_ next: [TileId]) {
        if let data = try? JSONSerialization.data(withJSONObject: next.map(\.rawValue)),
           let json = String(data: data, encoding: .utf8) {
            defaults.set(json, forKey: tileOrderKey)
        }
        ids = next
    }

    private static func readIds(from defaults: UserDefaults) -> [TileId] {
        guard let raw = defaults.string(forKey: tileOrderKey),
              let data = raw.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [Any]
        else { return defaultTileIds }
        // Preserve the chosen ORDER; drop unknown ids and duplicates. An empty
        // saved list is a real state — the user removed every tile — and must
        // NOT fall back to the defaults, or a tile they deleted comes back on
        // the next launch.
        var seen = Set<TileId>()
        return array
            .compactMap { $0 as? String }
            .compactMap(TileId.init(rawValue:))
            .filter { seen.insert($0).inserted }
    }

    private static func readWidths(from defaults: UserDefaults) -> [TileId: TileWidth] {
        guard let raw = defaults.string(forKey: tileSizeKey),
              let data = raw.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        var out: [TileId: TileWidth] = [:]
        for (key, value) in object {
            guard let id = TileId(rawValue: key),
                  let entry = value as? [String: Any],
                  let w = entry["w"] as? String,
                  let width = TileWidth(rawValue: w)
            else { continue }
            out[id] = width
        }
        return out
    }
}

File: apps/android/app/src/main/java/com/sanvya/app/ui/dashboard/DashboardPrefs.kt
package com.sanvya.app.ui.dashboard

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONArray
import org.json.JSONObject
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * Which tiles are on the dashboard, in what order, and how wide each one is.
 *
 * A faithful port of `apps/web/src/dashboard.ts`, down to the two storage keys
 * (`dashboardTiles`, `dashboardTileSizes`) and the JSON shapes, so a saved
 * dashboard means the same thing wherever it is read. Same pattern as
 * `NavPrefs` — a Koin-injected SharedPreferences behind a StateFlow.
 *
 * **This is per-device and does not sync**, and that is web's behaviour too:
 * `dashboard.ts` writes to localStorage, not to a table. Putting the native
 * copy in the database "because we can" would make two devices disagree about
 * a preference the browser has always kept to itself.
 *
 * Web stores a **third** key, `dashboardTileSpans`, plus `TileSpan`,
 * `setTileSpan`, `useTileSpans`, `H_ROWS` and `isTileEnabled`. Every one of
 * them is exported and unimported — a first sizing attempt that the {w,h}
 * system replaced and nobody deleted. None of it is ported.
 */
object DashboardPrefs : KoinComponent {

    private const val PREFS_NAME = "sanvya_prefs"

    private val context: Context by inject()
    private val sharedPrefs: SharedPreferences by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    private val _ids: MutableStateFlow<List<TileId>> by lazy { MutableStateFlow(readIds()) }

    /** Enabled tiles, in the user's chosen order. */
    val ids: StateFlow<List<TileId>> get() = _ids

    private val _widths: MutableStateFlow<Map<TileId, TileWidth>> by lazy { MutableStateFlow(readWidths()) }

    /** Per-tile width overrides. A tile with no entry uses [defaultWidth]. */
    val widths: StateFlow<Map<TileId, TileWidth>> get() = _widths

    /**
     * The width a tile starts at.
     *
     * Web's grid is four columns wide and its default size is `md` (two of
     * them) — half the row. A phone renders one column whatever this says, so
     * the value only becomes visible on a tablet, which is exactly where a
     * wrong default would look worst.
     */
    val defaultWidth: TileWidth = TileWidth.MD

    fun widthOf(id: TileId): TileWidth = _widths.value[id] ?: defaultWidth

    /** Add or remove a tile. Adding appends, which keeps a custom order intact. */
    fun setEnabled(id: TileId, on: Boolean) {
        val current = _ids.value
        val next = when {
            on && id !in current -> current + id
            !on -> current.filterNot { it == id }
            else -> return
        }
        writeIds(next)
    }

    /** Persist an explicit order — drag, or the move-up/move-down buttons. */
    fun reorder(ordered: List<TileId>) = writeIds(ordered.distinct())

    /** Move one tile by one position. Returns false at either end. */
    fun move(id: TileId, delta: Int): Boolean {
        val current = _ids.value.toMutableList()
        val from = current.indexOf(id)
        val to = from + delta
        if (from < 0 || to < 0 || to >= current.size) return false
        current.removeAt(from)
        current.add(to, id)
        writeIds(current)
        return true
    }

    fun setWidth(id: TileId, width: TileWidth) {
        val next = _widths.value + (id to width)
        val json = JSONObject()
        // Web stores {w,h}. Only `w` is ever read back -- `H_ROWS` exists but
        // nothing uses it, and row height is measured from content instead. The
        // `h` is written so a dashboard saved on a phone still parses in the
        // browser, rather than tripping web's `isDim(s.h)` guard and being
        // dropped whole.
        next.forEach { (tile, w) ->
            json.put(tile.key, JSONObject().put("w", w.name.lowercase()).put("h", "md"))
        }
        sharedPrefs.edit().putString(TILE_SIZE_KEY, json.toString()).apply()
        _widths.value = next
    }

    private fun writeIds(ids: List<TileId>) {
        sharedPrefs.edit().putString(TILE_ORDER_KEY, JSONArray(ids.map { it.key }).toString()).apply()
        _ids.value = ids
    }

    private fun readIds(): List<TileId> {
        val raw = sharedPrefs.getString(TILE_ORDER_KEY, null) ?: return DEFAULT_TILE_IDS
        return try {
            val array = JSONArray(raw)
            // Preserve the chosen ORDER; drop unknown ids and duplicates. An
            // empty saved list is a real state -- the user removed every tile --
            // and must NOT fall back to the defaults, or a tile they deleted
            // comes back on the next launch.
            (0 until array.length())
                .mapNotNull { TileId.from(array.optString(it, null)) }
                .distinct()
        } catch (_: Exception) {
            DEFAULT_TILE_IDS
        }
    }

    private fun readWidths(): Map<TileId, TileWidth> {
        val raw = sharedPrefs.getString(TILE_SIZE_KEY, null) ?: return emptyMap()
        return try {
            val json = JSONObject(raw)
            buildMap {
                json.keys().forEach { key ->
                    val id = TileId.from(key) ?: return@forEach
                    val w = json.optJSONObject(key)?.optString("w") ?: return@forEach
                    TileWidth.entries.firstOrNull { it.name.equals(w, ignoreCase = true) }
                        ?.let { put(id, it) }
                }
            }
        } catch (_: Exception) {
            emptyMap()
        }
    }
}
